### PR TITLE
 Remove `AcceleratorConnector.num_nodes` 

### LINKED ISSUE #12082: Remove `AcceleratorConnector.num_nodes`
## Proposed refactor
Remove the property `AcceleratorConnector.num_nodes` and migrate its usage
https://github.com/PyTorchLightning/pytorch-lightning/blob/a8ee5cacb7b31064c8b73372ca944d43f7caadc6/pytorch_lightning/trainer/connectors/accelerator_connector.py#L777-L779

### Motivation

`AcceleratorConnector.num_nodes` was only used on `Trainer.num_nodes` and this property were not meant to be public. 
`Trainer.num_nodes` is preferred to return the property from its strategy using `getattr(Trainer.strategy, "num_nodes", 1)` compared to access its accelerator_connector. 
 
After the property removal `Trainer.num_nodes` is the alternative of accessing `num_nodes`

### Pitch
https://github.com/PyTorchLightning/pytorch-lightning/blob/a8ee5cacb7b31064c8b73372ca944d43f7caadc6/pytorch_lightning/trainer/trainer.py#L2015-L2017

```
@property
def num_nodes(self) -> int:
    getattr(self.strategy, "num_nodes", 1):
```

cc @justusschock @awaelchli @rohitgr7 @four4fish

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?

Fixes #12082 (Suggested alternatives listed in the issue)
Part of #11449: deprecate unused properties in accelerator_connector

### Does your PR introduce any breaking changes? If yes, please list them.

N/A

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [N/A] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [N/A] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -2016,7 +2016,11 @@ def should_rank_save_checkpoint(self) -> bool:
 
     @property
     def num_nodes(self) -> int:
-        return self._accelerator_connector.num_nodes
+        rank_zero_deprecation(
+            "`Trainer.num_nodes` was deprecated in v1.6 and will be removed in v1.8."
+            ' Please use `getattr(Trainer.strategy, "num_nodes", 1)` instead.'
+        )
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2022]: do we need to deprecate it off the trainer? given it's a constructor argument to the Trainer, this is very convenient for users.

fully agreed with removing the duplicated property from the accelerator connector though

Comment (DuYicong515) [reply pytorch_lightning/trainer/trainer.py]: Discussed on the issue, will not deprecate this off the trainer.

**Review by ananthsub — APPROVED** [review]

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -2059,7 +2059,6 @@ def training_step(self, batch, batch_idx):
         ({"strategy": None}, SingleDeviceStrategy, "single_device", CPUAccelerator, 0),
         ({"strategy": "dp"}, DDPStrategy, "ddp", CPUAccelerator, 0),
         ({"strategy": "ddp"}, DDPStrategy, "ddp", CPUAccelerator, 0),
-        ({"strategy": "ddp", "num_processes": 2}, DDPStrategy, "ddp", CPUAccelerator, 0),
```
Comment (DuYicong515) [inline tests/trainer/test_trainer.py:2062]: This test case is the same with L2075, remove duplicated test case. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/7e2f9fbad555242b0ceb2a24e5e4c004f0701bae/tests/trainer/test_trainer.py#L2075

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -2097,6 +2096,23 @@ def training_step(self, batch, batch_idx):
             2,
         ),
         ({"strategy": DDPShardedStrategy(), "gpus": 2}, DDPShardedStrategy, "ddp_sharded", GPUAccelerator, 2),
+        ({"strategy": "ddp2", "gpus": 2, "num_nodes": 2}, DDP2Strategy, "ddp2", GPUAccelerator, 2),
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:2099]: Thanks for the additional test cases. 
I think in general we should probably also add a separate test for setting the value on any strategy instance.

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Pull request merged**